lua/neotest/consumers/watch/watcher.lua

```diff
@@ -0,0 +1,199 @@
+    error("No symbols query for language: " .. lang)
+  end
+  if type(query) == "function" then
+    return query(root, content, path)
+  end
+  local parsed_query = lib.treesitter.normalise_query(lang, query)
+  local symbols = {}
+  for id, node in parsed_query:iter_captures(root, content) do
+    if parsed_query.captures[id] == "symbol" then
+      symbols[#symbols + 1] = { node:range() }
+    end
+  end
+  return symbols
+end
+
+---@param args neotest.consumers.watch.watcher.WatchArgs
+---@return string[] paths
+function Watcher:_get_linked_files(path, root_path, args)
+  local symbols = lib.subprocess.enabled()
+      and lib.subprocess.call(
+        [[require("neotest.consumers.watch.watcher")._parse_symbols]],
+        { path }
+      )
+    or self._parse_symbols(path)
+  local path_uri = vim.uri_from_fname(path)
+  local dependency_uris = {}
+  logger.debug("Getting symbol definitions for", path)
+  for _, range in ipairs(symbols) do
+    local _, defs = self.lsp_client.request.textDocument_definition({
+      position = { line = range[1], character = range[2] },
+      textDocument = { uri = path_uri },
+    })
+
+    if defs ~= nil and type(defs[1]) ~= "table" then
+      defs = { defs }
+    end
+
+    for _, def in ipairs(defs or {}) do
+      dependency_uris[def.uri or def.targetUri] = true
+    end
+  end
+  local paths = { path }
+  for uri in pairs(dependency_uris) do
+    local p = vim.uri_to_fname(uri)
+    if uri ~= path_uri and args.filter_path(p, root_path) then
+      paths[#paths + 1] = p
+    end
+  end
+  logger.debug("Found", #paths, "linked files for", path)
+  return paths
+end
+
+---@class neotest.consumers.watch.watcher.WatchArgs
+---@field filter_path fun(root: string, path: string): boolean
+
+---@paam tree neotest.Tree
+function Watcher:_files_in_tree(tree)
+  if tree:data().type ~= "dir" then
+    return { tree:data().path }
+  end
+  local paths = {}
+  for _, pos in
+    tree:iter({
+      continue = function(node)
+        return node:data().type == "dir"
+      end,
+    })
+  do
+    if pos.type == "file" then
+      paths[#paths + 1] = pos.path
+    end
+  end
+  return paths
+end
+
+---@param root string
+---@param paths string[]
+---@param args neotest.consumers.watch.watcher.WatchArgs
+function Watcher:_build_dependencies(root, paths, args, dependencies)
+  local count = 0
+  local worker = function()
+    while #paths > 0 do
+      local path = table.remove(paths)
+
+      if not dependencies[path] then
+        count = count + 1
+        dependencies[path] = {}
+        local path_results = self:_get_linked_files(path, root, args)
+        dependencies[path] = path_results
+
+        for _, p in ipairs(path_results) do
+          if not dependencies[p] then
+            paths[#paths + 1] = p
+          end
+        end
+      end
+    end
+  end
+  local num_workers = 4
+  local workers = {}
+  for _ = 1, num_workers do
+    workers[#workers + 1] = worker
+  end
+  nio.gather(workers)
+end
```

Comment:
Can you describe this function's work details?

I am trying to understand why after building workers then, the thing works.